ext/curl/tests/curl_upload_redirect.phpt

```diff
@@ -0,0 +1,65 @@
+--TEST--
+Test curl_exec() function with basic functionality 
+--CREDITS--
+Sebastian Deutsch <sebastian.deutsch@9elements.com>
+TestFest 2009 - AFUP - Jean-Marc Fontaine <jmf@durcommefaire.net>
+--SKIPIF--
+<?php 
+if (!extension_loaded("curl")) exit("skip curl extension not loaded");
+if (false === getenv('PHP_CURL_HTTP_REMOTE_SERVER'))  exit("skip PHP_CURL_HTTP_REMOTE_SERVER env variable is not defined");
+?>
+--FILE--
+<?php
+/* Prototype  : bool curl_exec(resource ch)
+ * Description: Perform a cURL session 
+ * Source code: ext/curl/interface.c
+ * Alias to functions: 
+ */
+
+  $host = getenv('PHP_CURL_HTTP_REMOTE_SERVER');
+
+  // start testing
+  echo "*** Testing curl_exec() : basic functionality ***\n";
+
+  $fp = fopen(__FILE__, 'r');
+  /**
+   * @param resource $ch     curl handle
+   * @param resource $fd     file descriptor - do not use
+   * @param int      $offset move to position in file
+   * @param int      $whence start to seek from ...
+   *
+   * @return string
+   */
+  function seek_callback($ch, $fd, $offset, $whence)
+  {
+    global $fp;
+    return fseek($fp, $offset, $whence) == 0;
+  }
+
+  $url = "{$host}/get.php?test=redirect_upload";
+  $ch = curl_init();
+
+  ob_start(); // start output buffering
+  curl_setopt($ch, CURLOPT_INFILE, $fp);
+  curl_setopt($ch, CURLOPT_UPLOAD, 1);
+  curl_setopt($ch, CURLOPT_URL, $url); //set the url we want to use
+  curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
+  curl_setopt($ch, CURLOPT_SEEKFUNCTION, 'seek_callback');
```

Comment:
@laruence 
The error when uploading data ( not only files, streams too ) happens when the server to upload to replies with a redirect. So basically the next server is contacted and the upload is restarted by resetting the seek position. As long as the stream can be reopened this could work, but I would not really rely on that. So the easiest is really to reposition the seek pointer inside the stream, which is more flexible.